### PR TITLE
Qa 438 qa 439

### PR DESCRIPTION
**Pull request by TregubovYury**

### CODE REVIEW

File: mos_tests/environment/os_actions.py [hunk 1]
```diff
@@ -666,6 +667,9 @@ def force_l3_reshedule(self, router_id, new_l3_agt_id=''):
                                                  router_id)
         self.neutron.add_router_to_l3_agent(new_l3_agt_id,
                                             {"router_id": router_id})
+
         assert(wait(
```
Comment (gdyuldin) [inline mos_tests/environment/os_actions.py:671]: assert is redundant here

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 2]
```diff
@@ -37,6 +39,10 @@ def variables(self, init):
         self.zone = self.os_conn.nova.availability_zones.find(zoneName="nova")
         self.security_group = self.os_conn.create_sec_group_for_ssh()
         self.instance_keypair = self.os_conn.create_key(key_name='instancekey')
+        exist_networks = self.os_conn.list_networks()['networks']
+        self.ext_network = [x for x in exist_networks
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:43]: `self.os_conn.ext_network`

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 3]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
         self.check_ping_from_vm(vm=self.server2,
                                 vm_keypair=self.instance_keypair,
                                 ip_to_ping=self.server1_ip)
+
+    def try_to_update_router_to_centralized(self, router_id):
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:263]: This is check, so let's call it like `check_exception_on_router_update_to_centralize`

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 4]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+                                           {'router': {
+                                            'admin_state_up': False}})
+
+        with pytest.raises(NeutronClientException) as e:
+            self.os_conn.neutron.update_router(router_id,
+                                           {'router': {
+                                            'distributed': False}})
+            err_msg = 'Distributed parameter was updated, but in case of'
+            err_msg = err_msg + ' dvr feature that should not be possible!'
+            assert False, err_msg
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:276]: `pytest.raises` context manager will fail if there is no raised exception inside it

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 5]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+                    id=router_id)['routers'][0]
+        err_msg = 'distributed parameter for the router {0} is {1}.'.format(
+                   router['name'], router['distributed'])
+        err_msg = err_msg + " But it's expected value is True"
+        assert router['distributed'], err_msg
+
+        self.try_to_update_router_to_centralized(router['id'])
+
+    @pytest.fixture
+    def create_legacy_router(self, variables):
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:315]: let's rename to 'legacy_router' and return created router

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 6]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+
+        router = self.os_conn.create_router(
+                     name="router01", distributed=False)['router']
+        self.router_id = router['id']
+        self.os_conn.router_gateway_add(router_id=self.router_id,
+                                        network_id=self.ext_network['id'])
+        logger.info('router {} was created'.format(self.router_id))
+
+    @pytest.fixture
+    def create_dvr_router(self, variables):
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:330]: as above

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 7]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+                                           {'router': {
+                                            'distributed': True}})
+
+        # Change admin_state_up to True to enable the router
+        self.os_conn.neutron.update_router(self.router_id,
+                                           {'router': {
+                                            'admin_state_up': True}})
+
+        # Check that distributed is really changed to True
+        router = self.os_conn.neutron.list_routers(
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:428]: Does neutron client has `get_router` method?

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 8]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+                    id=self.router_id)['routers'][0]
+        err_msg = 'distributed parameter for the router {0} is {1}.'.format(
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:430]:
```
err_msg = ('distributed parameter for the router {0} is {1}. '
           'But it's expected value is True'
           ).format(router['name'], router['distributed'])
```

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 9]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+                   router['name'], router['distributed'])
+        err_msg = err_msg + " But it's expected value is True"
+        assert router['distributed'], err_msg
+
+        # Check that the router namespace is available on the compute now
+        with self.env.get_ssh_to_node(self.compute_ips[0]) as remote:
+            cmd = "ip netns | grep [q]router-{}".format(self.router_id)
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:437]: grep returns non-zero exit code if nothing finded.
You can use `remote.check_call(cmd)`

File: mos_tests/neutron/python_tests/test_dvr.py [hunk 10]
```diff
@@ -253,3 +259,344 @@ def test_routing_after_reset_computes(self, prepare_openstack, env_name):
...
+
+    def get_snat_controller(self, host_name):
+        snat_controller = None
+        controllers = self.env.get_nodes_by_role('controller')
+        for node in controllers:
+            if node.data['fqdn'] == host_name:
+                with self.env.get_ssh_to_node(node.data['ip']) as remote:
+                    cmd = 'ip netns | grep [s]nat-{}'.format(self.router_id)
+                    result = remote.execute(cmd)
+                    if 'snat' in result['stdout']:
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_dvr.py:458]: `result['stdout']` is list.
Is your code work as you expect?

File: mos_tests/neutron/python_tests/test_dvr.py (hunk 7 above)
Comment (TregubovYury) [inline mos_tests/neutron/python_tests/test_dvr.py:428]: No, only list_routers with filter is possible

File: mos_tests/neutron/python_tests/test_dvr.py (hunk 10 above)
Comment (TregubovYury) [inline mos_tests/neutron/python_tests/test_dvr.py:458]: Indeed, here should be another code: if any('snat' in s for s in result['stdout']). Crap... lost it somewhere...

**Comment by TregubovYury** [comment]
After applying of comments from pull#96 need to re-base with fresh master. will create new pull request to keep the history clear.

**Pull request closed without merging**